File: Python/csv_reader.py
import csv
from typing import List

from book import Book


class CSVReader:
    def __init__(self) -> None:
        self._books: List[Book] = []

    @property
    def get_books(self) -> List[Book]:
        return self._books

    def read_books_from_csv(self, csv_file_path: str) -> None:

        with open(csv_file_path, 'r', newline='') as csvfile:
            reader = csv.reader(csvfile)
            next(reader)  # skip header

            for row in reader:
                if row[0]:
                    try:
                        year = int(row[3])
                    except ValueError:
                        print(f"Invalid year: {row[3]}")
                        continue

                    book = Book(
                        isbn=row[0],
                        title=row[1],
                        author=row[2],
                        year=year)

                    self._books.append(book)

File: Go/utils/csv_reader.go
package utils

import (
	"encoding/csv"
	"os"
	"strconv"

	"github.com/siddhant-vij/Personal-Book-Library/models"
)

type CsvReader struct {
	books []*models.Book
}

func NewCsvReader() *CsvReader {
	return &CsvReader{
		books: make([]*models.Book, 0),
	}
}

func (c *CsvReader) GetBooks() []*models.Book {
	return c.books
}

func (c *CsvReader) ReadBooksFromCSV(csvFilePath string) error {
	file, err := os.Open(csvFilePath)
	if err != nil {
		return err
	}
	defer file.Close()

	csvReader := csv.NewReader(file)

	records, err := csvReader.ReadAll()
	if err != nil {
		return err
	}

	for _, record := range records {
		if len(record) == 0 || record[0] == "isbn" {
			continue
		}
		isbn := record[0]
		title := record[1]
		author := record[2]
		publicationYear, _ := strconv.Atoi(record[3])

		book := models.NewBook(isbn, title, author, publicationYear)

		c.books = append(c.books, book)
	}

	return nil
}
